File: src/rekognition.py
from pathlib import Path

import boto3
from botocore.exceptions import ClientError, NoCredentialsError

from .config import DEFAULT_CONFIG
from .models import DetectedInstance, DetectedLabel, DetectionResult


MAX_IMAGE_BYTES = 5 * 1024 * 1024


class RekognitionError(Exception):
    pass


class CredentialsError(RekognitionError):
    pass


class ImageTooLargeError(RekognitionError):
    pass


class InvalidImageError(RekognitionError):
    pass


class RekognitionClient:
    def __init__(
        self,
        region_name: str | None = None,
        min_confidence: float | None = None,
        max_labels: int | None = None,
    ):
        config = DEFAULT_CONFIG.rekognition
        self.region_name = region_name or config.region_name
        self.min_confidence = min_confidence or config.min_confidence
        self.max_labels = max_labels or config.max_labels

        try:
            self.client = boto3.client("rekognition", region_name=self.region_name)
        except NoCredentialsError:
            raise CredentialsError(
                "AWS credentials not found. Configure via environment variables, "
                "~/.aws/credentials, or IAM role."
            )

    def detect_labels(
        self,
        image_bytes: bytes,
        min_confidence: float | None = None,
        max_labels: int | None = None,
    ) -> DetectionResult:
        if len(image_bytes) > MAX_IMAGE_BYTES:
            raise ImageTooLargeError(
                f"Image size ({len(image_bytes)} bytes) exceeds Rekognition limit "
                f"of {MAX_IMAGE_BYTES} bytes (5MB). Resize or compress the image."
            )

        try:
            response = self.client.detect_labels(
                Image={"Bytes": image_bytes},
                MinConfidence=min_confidence or self.min_confidence,
                MaxLabels=max_labels or self.max_labels,
                Features=["GENERAL_LABELS"],
            )
        except ClientError as e:
            error_code = e.response["Error"]["Code"]
            if error_code == "InvalidImageFormatException":
                raise InvalidImageError("Image format not supported by Rekognition.")
            if error_code == "ImageTooLargeException":
                raise ImageTooLargeError("Image too large for Rekognition.")
            raise RekognitionError(f"Rekognition API error: {e}")
        except NoCredentialsError:
            raise CredentialsError("AWS credentials expired or invalid.")

        labels = self._parse_labels(response)

        return DetectionResult(
            labels=labels,
            raw_response=response,
            image_width=0,
            image_height=0,
        )

    def detect_labels_from_path(
        self,
        image_path: Path,
        min_confidence: float | None = None,
        max_labels: int | None = None,
    ) -> DetectionResult:
        image_path = Path(image_path)
        if not image_path.exists():
            raise FileNotFoundError(f"Image not found: {image_path}")

        image_bytes = image_path.read_bytes()
        return self.detect_labels(image_bytes, min_confidence, max_labels)

    def _parse_labels(self, response: dict) -> list[DetectedLabel]:
        labels = []
        for label_data in response.get("Labels", []):
            instances = []
            for inst in label_data.get("Instances", []):
                bbox = inst.get("BoundingBox", {})
                instances.append(
                    DetectedInstance(
                        bbox_left=bbox.get("Left", 0),
                        bbox_top=bbox.get("Top", 0),
                        bbox_width=bbox.get("Width", 0),
                        bbox_height=bbox.get("Height", 0),
                        confidence=inst.get("Confidence", 0),
                    )
                )

            categories = [
                c.get("Name", "") for c in label_data.get("Categories", [])
            ]
            parents = [
                p.get("Name", "") for p in label_data.get("Parents", [])
            ]

            labels.append(
                DetectedLabel(
                    name=label_data.get("Name", ""),
                    confidence=label_data.get("Confidence", 0),
                    categories=categories,
                    parents=parents,
                    instances=instances,
                )
            )

        return labels

File: src/lib/rekognition.ts
import {
  RekognitionClient,
  DetectLabelsCommand,
  type DetectLabelsCommandOutput,
  type Label,
  type Instance,
} from "@aws-sdk/client-rekognition";

import { rekognitionLogger as logger } from "./logger";
import { uploadToS3, getS3Bucket } from "./s3";

const REGION = process.env.AWS_REGION || "ap-northeast-1";
const MAX_INLINE_BYTES = 5 * 1024 * 1024; // 5MB

const client = new RekognitionClient({ region: REGION });

export class RekognitionError extends Error {
  code: string;

  constructor(message: string, code: string) {
    super(message);
    this.name = "RekognitionError";
    this.code = code;
  }
}

export interface DetectedInstance {
  confidence: number;
  boundingBox?: {
    left: number;
    top: number;
    width: number;
    height: number;
  };
}

export interface DetectedLabel {
  name: string;
  confidence: number;
  categories: string[];
  parents: string[];
  instances: DetectedInstance[];
}

export interface DetectionResult {
  labels: DetectedLabel[];
  labelCount: number;
  rawResponse: DetectLabelsCommandOutput;
}

function parseLabel(label: Label): DetectedLabel {
  const instances: DetectedInstance[] = (label.Instances || []).map(
    (inst: Instance) => ({
      confidence: inst.Confidence || 0,
      boundingBox: inst.BoundingBox
        ? {
            left: inst.BoundingBox.Left || 0,
            top: inst.BoundingBox.Top || 0,
            width: inst.BoundingBox.Width || 0,
            height: inst.BoundingBox.Height || 0,
          }
        : undefined,
    }),
  );

  return {
    name: label.Name || "Unknown",
    confidence: label.Confidence || 0,
    categories: (label.Categories || []).map((c) => c.Name || ""),
    parents: (label.Parents || []).map((p) => p.Name || ""),
    instances,
  };
}

export async function detectLabels(
  imageBytes: Buffer,
  options?: {
    minConfidence?: number;
    maxLabels?: number;
    s3Key?: string;
  },
): Promise<DetectionResult> {
  const minConfidence = options?.minConfidence ?? 50;
  const maxLabels = options?.maxLabels ?? 50;

  logger.info("Starting label detection", {
    imageSizeBytes: imageBytes.length,
    minConfidence,
    maxLabels,
  });

  let imageParam: { Bytes: Buffer } | { S3Object: { Bucket: string; Name: string } };

  if (imageBytes.length > MAX_INLINE_BYTES) {
    const bucket = getS3Bucket();
    if (!bucket) {
      throw new RekognitionError(
        `Image exceeds ${MAX_INLINE_BYTES} bytes and S3 is not configured for fallback`,
        "IMAGE_TOO_LARGE",
      );
    }

    const s3Key =
      options?.s3Key || `rekognition-tmp/${Date.now()}_oversized`;
    await uploadToS3(s3Key, imageBytes, "application/octet-stream");
    imageParam = { S3Object: { Bucket: bucket, Name: s3Key } };
    logger.info("Image exceeds inline limit, using S3 reference", { s3Key });
  } else {
    imageParam = { Bytes: imageBytes };
  }

  try {
    const command = new DetectLabelsCommand({
      Image: imageParam,
      MinConfidence: minConfidence,
      MaxLabels: maxLabels,
    });

    const response = await client.send(command);
    const labels = (response.Labels || []).map(parseLabel);

    logger.info("Detection completed", { labelCount: labels.length });

    return {
      labels,
      labelCount: labels.length,
      rawResponse: response,
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger.error("Rekognition API error", { error: { message } });
    throw new RekognitionError(message, "REKOGNITION_API_ERROR");
  }
}
